ibc-testkit/src/testapp/ibc/core/core_ctx.rs

```diff
@@ -298,46 +268,449 @@ impl ValidationContext for MockContext {
+                    .get(StoreHeight::Pending, commitment_path)
+                    .is_some()
+            })
+            .map(|commitment_path| commitment_path.sequence)
+            .collect())
+    }
+}
+
+impl<S> ExecutionContext for MockIbcStore<S>
+where
+    S: ProvableStore + Debug,
+{
     type E = Self;
 
     fn get_client_execution_context(&mut self) -> &mut Self::E {
         self
     }
 
+    /// Called upon client creation.
+    /// Increases the counter which keeps track of how many clients have been created.
+    /// Should never fail.
     fn increase_client_counter(&mut self) -> Result<(), ContextError> {
-        self.ibc_store.lock().client_ids_counter += 1;
+        let current_sequence = self
+            .client_counter
+            .get(StoreHeight::Pending, &NextClientSequencePath)
+            .ok_or(ClientError::Other {
+                description: "client counter not found".into(),
+            })?;
+
+        self.client_counter
+            .set(NextClientSequencePath, current_sequence + 1)
+            .map_err(|e| ClientError::Other {
+                description: format!("client counter update failed: {e:?}"),
+            })?;
+
         Ok(())
     }
 
+    /// Stores the given connection_end at path
     fn store_connection(
         &mut self,
         connection_path: &ConnectionPath,
         connection_end: ConnectionEnd,
     ) -> Result<(), ContextError> {
-        let connection_id = connection_path.0.clone();
-        self.ibc_store
-            .lock()
-            .connections
-            .insert(connection_id, connection_end);
+        self.connection_end_store
+            .set(connection_path.clone(), connection_end)
+            .map_err(|_| ConnectionError::Other {
+                description: "Connection end store error".to_string(),
+            })?;
         Ok(())
     }
 
+    /// Stores the given connection_id at a path associated with the client_id.
     fn store_connection_to_client(
         &mut self,
         client_connection_path: &ClientConnectionPath,
         conn_id: ConnectionId,
     ) -> Result<(), ContextError> {
-        let client_id = client_connection_path.0.clone();
-        self.ibc_store
-            .lock()
-            .client_connections
-            .insert(client_id, conn_id);
+        let mut conn_ids: Vec<ConnectionId> = self
+            .connection_ids_store
+            .get(StoreHeight::Pending, client_connection_path)
+            .unwrap_or_default();
+        conn_ids.push(conn_id);
+        self.connection_ids_store
+            .set(client_connection_path.clone(), conn_ids)
+            .map_err(|_| ConnectionError::Other {
+                description: "Connection ids store error".to_string(),
+            })?;
         Ok(())
     }
 
+    /// Called upon connection identifier creation (Init or Try process).
+    /// Increases the counter which keeps track of how many connections have been created.
+    /// Should never fail.
```

Comment:
Similar point here with the "Should never fail" comment.

Reply:
d8fe3d4b